File: LeetCode/Problems/00003. Longest Substring Without Repeating Characters/Solution.cpp
#include <iostream>
#include <string> 
#include <set>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int max = 0;
        set<char> set;
        
        for (int i=0; i < s.size() ; i++){
            if ( (i+max) > s.size() ) break;

            set.clear();
            for (int j=i; j<(i+max); j++) set.insert( s[j] );
            
            if (max == set.size()){
                for (int j=(i+max); j<s.size(); j++ ){
                    auto pos = set.find( s[j] );
                    if(pos == set.end()){
                        set.insert(s[j]);
                        max++;
                    }else{
                        break;
                    }
                }
            }
        }
        
        return max;
    }
};


int main() {
    string s = "pwwkew";
    int output;

    Solution solution;
    output = solution.lengthOfLongestSubstring(s);

    cout << "output: " << output << endl;
    
    return 0;
}

File: LeetCode/Problems/00003. Longest Substring Without Repeating Characters/Solution.py

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        _max = 0
        _set = set()
        
        for i in range(len(s)):
            if (i + _max) > len(s):
                break
                
            _set = set(s[i:i+_max])
            if _max == len(_set): 
                for c in s[i+_max:]:
                    if c not in _set:
                        _set.add(c)
                        _max += 1
                    else:
                        break
        
        
        return _max

if __name__ == '__main__':
    s = "abcabcbb"
    
    solution = Solution()
    output = solution.lengthOfLongestSubstring(s)
    print("Output: ", output)
    
